src/Hint/NewType.hs

```diff
@@ -26,39 +31,110 @@ newtype Foo = Foo Int deriving stock Show
 -}
 module Hint.NewType (newtypeHint) where
 
-import Hint.Type
+import Hint.Type (Idea, DeclHint', Note(DecreasesLaziness), ideaNote, ignoreNoSuggestion', suggestN')
 
-newtypeHint :: DeclHint
+import Data.List (isSuffixOf)
+import "ghc-lib-parser" HsDecls
+import "ghc-lib-parser" HsSyn
+import "ghc-lib-parser" Outputable
+import "ghc-lib-parser" SrcLoc
+
+newtypeHint :: DeclHint'
 newtypeHint _ _ x = newtypeHintDecl x ++ newTypeDerivingStrategiesHintDecl x
 
-newtypeHintDecl :: Decl_ -> [Idea]
-newtypeHintDecl x
-    | Just (DataType s, t, f) <- singleSimpleField x
-    = [(suggestN "Use newtype instead of data" x $ f (NewType s) $ fromTyBang t)
-            {ideaNote = [DecreasesLaziness | not $ isTyBang t]}]
+newtypeHintDecl :: LHsDecl GhcPs -> [Idea]
+newtypeHintDecl old
+    | Just WarnNewtype{newDecl, insideType} <- singleSimpleFieldNew old
+    = [(suggestN' "Use newtype instead of data" old newDecl)
+            {ideaNote = [DecreasesLaziness | warnBang insideType]}]
 newtypeHintDecl _ = []
 
+newTypeDerivingStrategiesHintDecl :: LHsDecl GhcPs -> [Idea]
+newTypeDerivingStrategiesHintDecl decl@(L _ (TyClD _ (DataDecl _ _ _ _ dataDef))) =
+    [ignoreNoSuggestion' "Use DerivingStrategies" decl | not $ isData dataDef, not $ hasAllStrategies dataDef]
+newTypeDerivingStrategiesHintDecl _ = []
+
+hasAllStrategies :: HsDataDefn GhcPs -> Bool
+hasAllStrategies (HsDataDefn _ NewType _ _ _ _ (L _ xs)) = all hasStrategyClause xs
+hasAllStrategies _ = False
+
+isData :: HsDataDefn GhcPs -> Bool
+isData (HsDataDefn _ NewType _ _ _ _ _) = False
+isData (HsDataDefn _ DataType _ _ _ _ _) = True
+isData _ = False
+
+hasStrategyClause :: LHsDerivingClause GhcPs -> Bool
+hasStrategyClause (L _ (HsDerivingClause _ (Just _) _)) = True
+hasStrategyClause _ = False
+
+data WarnNewtype = WarnNewtype
+    { newDecl :: LHsDecl GhcPs
+    , insideType :: HsType GhcPs
+    }
+
+-- | Given a declaration, returns the suggested \"newtype\"ized declaration following these guidelines:
```

Comment:
This comment could probably go to the file header comment. What do you think?

Reply:
Happy either way.